logging/logging.go

```diff
@@ -72,25 +96,76 @@ func GetLogLevel() int {
 	return LogDiscard
 }
 
-// CustomizeLogging changes logging format
-func CustomizeLogging(loggingFormat string, serviceName string) {
-	log.SetOutput(os.Stderr)
-	log.SetFormatter(logFormatterFor(loggingFormat, serviceName))
+var (
+	errNoWriter      = errors.New("output writer is not specified")
+	errNoServiceName = errors.New("service name is not specified")
+	errNoFormat      = errors.New("log format is not specified")
+)
 
-	log.Debugf("Changed logging format to %s", loggingFormat)
+type CustomizeBuilder struct {
+	writer        io.Writer
+	serviceName   string
+	loggingFormat string
+	hooks         []FormatterHook
 }
 
-func logFormatterFor(loggingFormat string, serviceName string) log.Formatter {
-	loggingFormat = strings.ToLower(loggingFormat)
+func Customize() *CustomizeBuilder {
+	return &CustomizeBuilder{}
+}
 
-	if loggingFormat == "json" {
-		return JSONFormatter(log.Fields{FieldKeyProduct: serviceName})
+func (c *CustomizeBuilder) SetOutput(w io.Writer) *CustomizeBuilder {
+	c.writer = w
+	return c
+}
 
-	} else if loggingFormat == "cef" {
-		return CEFFormatter(log.Fields{FieldKeyProduct: serviceName})
+func (c *CustomizeBuilder) SetServiceName(serviceName string) *CustomizeBuilder {
+	c.serviceName = serviceName
+	return c
+}
+
+func (c *CustomizeBuilder) SetFormat(loggingFormat string) *CustomizeBuilder {
+	c.loggingFormat = loggingFormat
+	return c
+}
+
+func (c *CustomizeBuilder) SetHooks(hooks []FormatterHook) *CustomizeBuilder {
+	c.hooks = hooks
+	return c
+}
+
+func (c *CustomizeBuilder) Complete() error {
+	if c.writer == nil {
+		return errNoWriter
+	}
+	if c.serviceName == "" {
+		return errNoServiceName
 	}
+	if c.loggingFormat == "" {
+		return errNoFormat
+	}
```

Comment:
I think it would be nice to provide some sensible defaults. Like, use plaintext format, write logs to console, do not set the service field, and have no custom hooks. Whatever we do without `Customize()`. That way if one writes

```go
logging.Customize().Complete()
```

then they will get exactly the same default log instead of an unexpected error.

Reply:
Yes. Nice suggestion